kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,849 @@
+//! Defines [`LogReplayScanner`] used by [`TableChangesScan`] to process commit files and extract
+//! the metadata needed to generate the Change Data Feed.
+
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, LazyLock};
+
+use crate::actions::visitors::{visit_deletion_vector_at, ProtocolVisitor};
+use crate::actions::{
+    get_log_add_schema, get_log_schema, Protocol, ADD_NAME, CDC_NAME, COMMIT_INFO_NAME,
+    METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME,
+};
+use crate::engine_data::TypedGetData;
+use crate::expressions::{column_expr, column_name, Expression};
+use crate::path::ParsedLogPath;
+use crate::scan::data_skipping::DataSkippingFilter;
+use crate::scan::scan_row_schema;
+use crate::scan::state::DvInfo;
+use crate::schema::{ArrayType, ColumnNamesAndTypes, DataType, MapType, SchemaRef, StructType};
+use crate::table_properties::TableProperties;
+use crate::utils::require;
+use crate::{
+    DeltaResult, Engine, EngineData, Error, ExpressionHandler, ExpressionRef, JsonHandler,
+    RowVisitor,
+};
+use itertools::Itertools;
+
+pub struct TableChangesScanData {
+    pub data: Box<dyn EngineData>,
+    // The selection vector used to filter log actions.
+    pub selection_vector: Vec<bool>,
+    // An optional map from a remove action's path to its deletion vector
+    pub remove_dv: Option<Arc<HashMap<String, DvInfo>>>,
+}
+
+/// Given an iterator of ParsedLogPath and a predicate, returns an iterator of
+/// [`TableChangesScanData`]  Each row that is selected in the returned
+/// `engine_data` _must_ be processed to complete the scan. Non-selected rows
+/// _must_ be ignored.
+pub(crate) fn table_changes_action_iter(
+    engine: &dyn Engine,
+    commit_files: impl IntoIterator<Item = ParsedLogPath>,
+    table_schema: SchemaRef,
+    predicate: Option<ExpressionRef>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+    let json_handler = engine.get_json_handler();
+    let filter = DataSkippingFilter::new(engine, &table_schema, predicate);
+
+    let expression_handler = engine.get_expression_handler();
+    let result = commit_files
+        .into_iter()
+        .map(move |commit_file| -> DeltaResult<_> {
+            let mut scanner = LogReplayScanner::new(
```

Comment:
would it make sense to pass the engine and the predicate to `new`? Then you can only have to clone the `predicate` which is an `Arc` anyway. This is similar to how we do it in the `scan::LogReplayScanner` one.

Then any methods that need a handler just also take an `engine` as a arg, so they can get at it, like `process_scan_batch` in `scan::LogReplayScanner`.

This will really simplify ever using this over FFI

Reply:
Aha so that we don't have to expose the concept of a `DataSkippingFilter` to the engine. Makes sense

Reply:
update: we figured out the FFI surface is anyway a level above this code, so it doesn't matter.